packages/optimistic-oracle/test/proposer.js

```diff
@@ -0,0 +1,312 @@
+
+      // Submit all of the proposals now for all pending price requests.
+      await proposer.sendProposals();
+
+      // There should be one call for each unique collateral currency and one for each price request.
+      const expectedCallCount = collateralCurrenciesForIdentifier.length + identifiersToTest.length + 1;
+      assert.equal(expectedCallCount, spy.callCount);
+
+      // Check that only 1 price feed is cached for each unique identifier.
+      assert.equal(Object.keys(proposer.priceFeedCache).length, identifiersToTest.length);
+    });
+  });
+
+  it("Skip price requests with identifiers that proposer cannot construct a price feed for", async function() {
+    // Request a valid identifier but set an invalid price feed config:
+    await optimisticRequester.requestPrice(
+      identifiersToTest[0],
+      requestTime,
+      "0x",
+      collateralCurrenciesForIdentifier[0].address,
+      0
+    );
+
+    // This pricefeed config will cause the proposer to fail to construct a price feed because the
+    // PriceFeedMock type requires a `currentPrice` and a `historicalPrice` to be specified, which are missing
+    // here (and also not included in the DefaultPriceFeedConfig for the tested identifiers).
+    let invalidPriceFeedConfig = {};
+    proposer = new OptimisticOracleProposer({
+      logger: spyLogger,
+      optimisticOracleClient: client,
+      gasEstimator,
+      account: botRunner,
+      defaultPriceFeedConfig: invalidPriceFeedConfig
+    });
+    await proposer.update();
+    await proposer.sendProposals();
+
+    // Check for the specific failure log emitted by the proposer.
+    assert.equal(lastSpyLogLevel(spy), "error");
+    assert.isTrue(spyLogIncludes(spy, -1, "Failed to construct a PriceFeed for price request"));
+  });
+
+  it("Skip price requests with historical prices that proposer fails to fetch", async function() {
+    // Request a valid identifier that is getting bad data from the data source.
+    // Note: "INVALID" maps specifically to the InvalidPriceFeedMock in the DefaultPriceFeedConfig.js file.
+    const invalidPriceFeedIdentifier = web3.utils.utf8ToHex("INVALID");
+    await identifierWhitelist.addSupportedIdentifier(invalidPriceFeedIdentifier);
+    await optimisticRequester.requestPrice(
+      invalidPriceFeedIdentifier,
+      requestTime,
+      "0x",
+      collateralCurrenciesForIdentifier[0].address,
+      // collateral token doesn't matter as the error should log before its used
+      0
+    );
+    proposer = new OptimisticOracleProposer({
+      logger: spyLogger,
+      optimisticOracleClient: client,
+      gasEstimator,
+      account: botRunner
+    });
+    await proposer.update();
+    await proposer.sendProposals();
+
+    // Check for the specific failure log emitted by the proposer.
+    assert.equal(lastSpyLogLevel(spy), "error");
+    assert.isTrue(spyLogIncludes(spy, -1, "Failed to query historical price for price request"));
+  });
+
+  // TODO:
```

Comment:
```suggestion
  // TODO:
  // Can deal with ancillary data
```

should also have a test to validate ancillary data is correctly dealt with.

Reply:
Ah yes. Can you look at this [commit](https://github.com/UMAprotocol/protocol/pull/2505/commits/8042aa6c391756e9119fc10517f249c930ed24b5) and check that I'm correctly simulating how the Perpetual and EMP will submit ancillary data to the OO?